hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/audit/S3AAuditLogMergerAndParser.java

```diff
@@ -0,0 +1,546 @@
+          try {
+            if (longValues.contains(key)) {
+              if (value.equals("-")) {
+                avroDataRecord.put(key, null);
+              } else {
+                avroDataRecord.put(key, Long.parseLong(value));
+              }
+            } else {
+              avroDataRecord.put(key, value);
+            }
+          } catch (Exception e) {
+            avroDataRecord.put(key, null);
+          }
+        }
+        avroDataRecord.put("referrerMap", referrerHeaderList.get(count));
+        dataFileWriter.append(avroDataRecord);
+        byte[] byteArray = avroDataRecord.toString().getBytes("UTF-8");
+        fsDataOutputStream.write(byteArray);
+        count += 1;
+      }
+    }
+    dataFileWriter.close();
+    LOG.info("Successfully generated avro data");
+  }
+
+  /**
+   * Merge and parse all the audit log files and convert data into avro file.
+   *
+   * @param s3AFileSystem filesystem
+   * @param s3LogsPath    source path of logs
+   * @param s3DestPath    destination path of merged log file
+   * @return true
+   * @throws IOException on any failure
+   */
+  public boolean mergeAndParseAuditLogFiles(S3AFileSystem s3AFileSystem,
+      Path s3LogsPath,
+      Path s3DestPath) throws IOException {
+
+    RemoteIterator<LocatedFileStatus> listOfS3LogFiles =
+        s3AFileSystem.listFiles(s3LogsPath, true);
+
+    Path destFile = new Path(s3DestPath, "AuditLogFile");
+
+    List<HashMap<String, String>> entireAuditLogList = new ArrayList<>();
+    List<HashMap<String, String>> referrerHeaderList = new ArrayList<>();
+    List<HashMap<String, String>> auditLogList = new ArrayList<>();
+
+    try (FSDataOutputStream fsDataOutputStream = s3AFileSystem.create(
+        destFile)) {
+      // Iterating over the list of files to merge and parse
+      while (listOfS3LogFiles.hasNext()) {
+        Path logFile = listOfS3LogFiles.next().getPath();
+        FileStatus fileStatus = s3AFileSystem.getFileStatus(logFile);
+        int fileLength = (int) fileStatus.getLen();
+        byte[] byteBuffer = new byte[fileLength];
+        try (
+            FSDataInputStream fsDataInputStream = s3AFileSystem.open(logFile)) {
+
+          // Reads the file data in byte array
+          fsDataInputStream.readFully(byteBuffer);
+
+          // Convert byte array to string to make parsing easy
+          String byteStr = new String(byteBuffer, "UTF-8");
+          List<String> logs = IOUtils.readLines(new StringReader(byteStr));
+
+          // Parse each and every audit log from list of logs
+          for (String singleAuditLog : logs) {
+            // Parse audit log except referrer header
+            Map<String, String> auditLogMap = parseAuditLog(singleAuditLog);
+
+            String referrerHeader = auditLogMap.get("referrer");
+            if (referrerHeader == null || referrerHeader.equals("-")) {
+              LOG.info("Log didn't parsed : " + referrerHeader);
```

Comment:
nit, log at debug ahd use {} for insertion into the log

Reply:
modified to debug